### PR TITLE
Add in type-to-filter attributes option for Import from CSV

### PR DESCRIPTION
**Pull request by CrucisGamma**
### Description of the Change

Modify the "Import CSV Dialog" to be able to filter the many attributes by typing in a partial match on a number of attributes names.

### Alternate Designs

I did have an extra menu item under the existing 'select all' checkbox but it felt clunky to use.  Simply typing in 'Lati' to get all the Latitude fields or "Geo" to get geo or Date to get the date ones feels a lot more intuitive.

### Why Should This Be In Core?

Looking and pecking through the 140+ 'All' fields feels exceedingly cumbersome to find the field you're interested in.

### Benefits

People (I) won't be frustrated by finding the one attribute they actually want to use

### Possible Drawbacks

A Corona pandemic will occur in early-mid 2020... Does anybody read this?  I don't think there are any.

### Verification Process

- How did you verify that all new functionality works as expected?
Have run it and tested the new functionality without any appreciable differences in performance.  Imported and used a sample CSV file of my choosing.

- How did you verify that all changed functionality works as expected?
Have reviewed the changes to the code to hopefully ensure nothing to outrageous is going to go in.  Have done 'Fix imports' and 'Format' on all files to ensure correctly formatted code.

- How did you verify that the change has not introduced any regressions?
Rebuilt code.

Basically Import -> Delimited file , and then started typing in things I wanted to filter on (geo, lat, long, datetime)

### Applicable Issues

Nil.

### CODE REVIEW

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java [hunk 1]
```diff
@@ -488,6 +508,10 @@ public void setShowAllSchemaAttributes(final boolean showAllSchemaAttributes) {
         // TODO: the tick box could have changed but the menu item isn't updated, fix it
     }
 
+    public void setShowFilteredSchemaAttributes(final String filterAttributes) {
```
Comment (cygnus-x-1) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java:511]: This should be `setAttributeFilter`, and the variable should be `attributeFilter`.

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java [hunk 2]
```diff
@@ -328,8 +334,22 @@ public void updateDisplayedAttributes() {
 
     private Map<String, Attribute> createDisplayedAttributes(final Map<String, Attribute> autoAddedAttributes, final Map<String, Attribute> manuallyAddedAttributes) {
         Map<String, Attribute> displayedAttributes = new HashMap<>();
-        displayedAttributes.putAll(manuallyAddedAttributes);
-        displayedAttributes.putAll(autoAddedAttributes);
+
+        if (filterAttributes != null &&  filterAttributes.length() > 0) {
+            for (String name : autoAddedAttributes.keySet()) {
+                if (name.toUpperCase().matches(".*" + filterAttributes.toUpperCase() + ".*")) {
```
Comment (cygnus-x-1) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java:340]: You should use `contains` over `matches` with a regex.

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java [hunk 3]
```diff
@@ -328,8 +334,22 @@ public void updateDisplayedAttributes() {
 
     private Map<String, Attribute> createDisplayedAttributes(final Map<String, Attribute> autoAddedAttributes, final Map<String, Attribute> manuallyAddedAttributes) {
         Map<String, Attribute> displayedAttributes = new HashMap<>();
-        displayedAttributes.putAll(manuallyAddedAttributes);
-        displayedAttributes.putAll(autoAddedAttributes);
+
+        if (filterAttributes != null &&  filterAttributes.length() > 0) {
```
Comment (cygnus-x-1) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java:338]: I think we have a utility method to check for both bull and blank - @arcturus2?

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java [hunk 4]
```diff
@@ -328,8 +334,22 @@ public void updateDisplayedAttributes() {
...
+            for (String name : autoAddedAttributes.keySet()) {
```
Comment (cygnus-x-1) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java:339]: Make name final.

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java [hunk 5]
```diff
@@ -328,8 +334,22 @@ public void updateDisplayedAttributes() {
...
+                if (name.toUpperCase().matches(".*" + filterAttributes.toUpperCase() + ".*")) {
+                    displayedAttributes.put(name, autoAddedAttributes.get(name));
+                }
+            }
+            for (String name : manuallyAddedAttributes.keySet()) {
```
Comment (cygnus-x-1) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java:344]: Make name final.

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java [hunk 6]
```diff
@@ -170,7 +171,12 @@ public void setClearManuallyAdded(final boolean b) {
     }
 
     public void setDestination(final ImportDestination<?> destination) {
-        this.currentDestination = destination;
+        if (destination != null) {
```
Comment (cygnus-x-1) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java:174]: You could make this more concise using:
```
if (destination == null) {
    return;
}
this.currentDestination = destination;
```
...but is there an issue with setting `currentDestination` to null?

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/RunPane.java [hunk 7]
```diff
@@ -186,6 +189,19 @@ public RunPane(final ImportController importController) {
 
         attributePane.addRow(0, sourceVertexScrollPane, destinationVertexScrollPane, transactionScrollPane);
 
+        attributePane.setOnKeyPressed(event -> {
```
Comment (cygnus-x-1) [inline CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/RunPane.java:192]: In the interest of good UX, I would strongly recommend having a visual element to this filter. Having it just hide attributes as you type with no visual feedback is begging for erroneous bug reports.

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java (hunk 6 above)
Comment (CrucisGamma) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java:174]: No, I specifically am calling addDestination with 'null' from the RunPane.java call.  If it's already got a valid destination setup, I don't want to overwrite it.  I kept with the 'this.currentDestination' which should probably be just 'currentDestination' though.

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java (hunk 3 above)
Comment (CrucisGamma) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java:338]: isEmpty?

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/RunPane.java (hunk 7 above)
Comment (CrucisGamma) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/RunPane.java:192]: I've added a blurb to the 'help text' on that middle section and the responsiveness of it changing is fairly clear that you caused something to change when you start typing.  I did have it as a text edit button under Options, but that sucked

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java (hunk 3 above)
Comment (antares1470) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java:338]: StringUtils.isBlank(...) and StringUtils.isNotBlank(...)

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java (hunk 6 above)
Comment (antares1470) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java:174]: Sonar will flag the return statement in the void function so I'd say it would be ideal if it can be written without it (if there is a nice way of doing so)

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/RunPane.java (hunk 7 above)
Comment (cygnus-x-1) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/RunPane.java:192]: What about something like, when you start typing, a text input pops up to show that you're doing something?

**Comment by arcturus2** [comment]
>A Corona pandemic will occur in early-mid 2020... Does anybody read this? I don't think there are any.

I read it @CrucisGamma ! Note that this is a generic template so if something isn't applicable then you can leave that section blank. I've got an idea to improve some of the sections in this template so stay tuned :smiley:

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java (hunk 3 above)
Comment (arcturus2) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java:338]: Sorry for the delay @cygnus-x-1 in response, @antares1470 thanks for chipping in!

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java (hunk 6 above)
Comment (arcturus2) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java:174]: @CrucisGamma @antares1470  I've tagged you both on a different PR but I did a test to see if SQ complained about an empty return and it was OK with it. See #401.

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/RunPane.java (hunk 7 above)
Comment (CrucisGamma) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/RunPane.java:193]: Have added the text field in.

File: CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java (hunk 3 above)
Comment (CrucisGamma) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java]: I can find only one instance of 'isBlank' being called (at , 
CoreUtilities\src\au\gov\asd\tac\constellation\utilities\temporal\TemporalFormatting.java line 497ish), no isNotBlank and about 515 isEmpty (515 empties? We are a territory but this is not NT).

Comment (antares1470) [reply CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportController.java]: StringUtils.isBlank(str) is equivalent to str == null || str.IsEmpty().
Similarly, StringUtils.isNotBlank(str) is equivalent to str !=null && !str.isEmpty()

We probably haven't gotten around to fixing all instances yet but this is the way we'd prefer to test for emptiness (unless of course you know it can't be null in which case just use isEmpty)

**Comment by cygnus-x-1** [comment]
@CrucisGamma you will need to merge master into this branch and resolve the merge conflicts before we can resolve this pull request.
